### PR TITLE
feat: Add basic regions support to G4 example

### PR DESCRIPTION
**Pull request by ntadej**
I work with fast simulation for silicon detectors and I was interested to see how `G4ProductionCuts` affect the number of secondary particles. This may be useful also for other people so adding it here.

/cc @andiwand @paulgessinger

### CODE REVIEW

File: Examples/Algorithms/Geant4/include/ActsExamples/Geant4/RegionCreator.hpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    double gammaCut{};
+
+    /// Cut to be applied for electrons
+    double electronCut{};
+
+    /// Cut to be applied for positrons
+    double positronCut{};
+
+    /// Cut to be applied for protons
+    double protonCut{};
```
Comment (andiwand) [inline Examples/Algorithms/Geant4/include/ActsExamples/Geant4/RegionCreator.hpp:33]: what kind of cuts are these? initial energy of the created particle?

File: Examples/Algorithms/Geant4/include/ActsExamples/Geant4/RegionCreator.hpp [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  };
+
+  /// Region creator constructor
+  ///
+  /// @param cfg is the configuration struct
+  /// @param name is the region name
+  /// @param level is the logging level to be used
+  RegionCreator(const Config& cfg, std::string name,
+                Acts::Logging::Level level);
+  ~RegionCreator() {}
```
Comment (andiwand) [inline Examples/Algorithms/Geant4/include/ActsExamples/Geant4/RegionCreator.hpp:46]:
```suggestion
```

I think the destructor definition can be dropped or use `= default`

File: Examples/Algorithms/Geant4/include/ActsExamples/DDG4/DDG4DetectorConstruction.hpp [hunk 3]
```diff
@@ -28,7 +29,9 @@ struct DD4hepDetector;
 /// Construct the Geant4 detector from a DD4hep description.
 class DDG4DetectorConstruction final : public G4VUserDetectorConstruction {
  public:
-  DDG4DetectorConstruction(std::shared_ptr<DD4hep::DD4hepDetector> detector);
+  DDG4DetectorConstruction(
+      std::shared_ptr<DD4hep::DD4hepDetector> detector,
+      std::vector<std::shared_ptr<RegionCreator>> regionCreators);
```
Comment (andiwand) [inline Examples/Algorithms/Geant4/include/ActsExamples/DDG4/DDG4DetectorConstruction.hpp:34]: I wonder if sharing these has a benefit or if we can just do `std::vector<RegionCreator>`

**Review by andiwand — COMMENTED** [review]
I never used this cut but I can imagine it be useful!

File: Examples/Algorithms/Geant4/include/ActsExamples/DDG4/DDG4DetectorConstruction.hpp (hunk 3 above)
Comment (ntadej) [reply Examples/Algorithms/Geant4/include/ActsExamples/DDG4/DDG4DetectorConstruction.hpp:34]: It has `std::unique_ptr` members so they are not trivially copiable. Other complex structures that come from python are also stored in shared pointers and I just followed the same conventions.

I'm also not sure what the lifetime relation between the factory and the wrapper class is to pass references/pointers around.

File: Examples/Algorithms/Geant4/include/ActsExamples/Geant4/RegionCreator.hpp (hunk 2 above)
Comment (ntadej) [reply Examples/Algorithms/Geant4/include/ActsExamples/Geant4/RegionCreator.hpp]: Done.

File: Examples/Algorithms/Geant4/include/ActsExamples/Geant4/RegionCreator.hpp (hunk 1 above)
Comment (ntadej) [reply Examples/Algorithms/Geant4/include/ActsExamples/Geant4/RegionCreator.hpp]: Yes, it's the cut on energy but in length units. That's why it is often set per detector type so it is optimised on the relevant material. I tried to improve the documentation.

**Review by andiwand — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results [d1511d004145a448c4e0de8d01f5ad03efcd6828]
### :white_check_mark: All tests successful

|status|job|report|
|------|---|------|
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294623)||
|:green_circle: | [test_ActsEFTrackFit](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294622)||
|:green_circle: | [test_ActsPersistifySeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294621)||
|:green_circle: | [test_ActsBenchmarkWithSpot](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294620)||
|:green_circle: | [test_ActsAnalogueClustering](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294619)||
|:green_circle: | [test_ActsConversionWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294618)||
|:green_circle: | [test_ActsWorkflowFastTracking](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294617)||
|:green_circle: | [test_ActsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294616)||
|:green_circle: | [test_ActsValidateAmbiguityResolution](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294615)||
|:green_circle: | [test_ActsValidateResolvedTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294614)||
|:green_circle: | [test_ActsValidateTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294613)||
|:green_circle: | [test_ActsValidateActsCoreSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294612)||
|:green_circle: | [test_ActsValidateActsSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294611)||
|:green_circle: | [test_ActsValidateSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294610)||
|:green_circle: | [test_ActsValidateOrthogonalSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294609)||
|:green_circle: | [test_ActsValidateClusters](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294608)||
|:green_circle: | [test_ActsPersistifyEDM](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294607)||
|:green_circle: | [test_ActsGSFRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294606)||
|:green_circle: | [test_ActsKfRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294605)||
|:green_circle: | [test_ActsExtrapolationAlgTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294604)||
|:green_circle: | [test_ActsITkTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294603)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294602)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294601)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294600)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294599)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294598)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294597)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38294596)||

**Pull request merged**